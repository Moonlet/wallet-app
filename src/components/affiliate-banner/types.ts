export const AffiliateBannerType = {
    LEDGER_NANO_X: {
        url: 'https://shop.ledger.com/pages/ledger-nano-x?r=69d534ac55d9',
        image: require('../../assets/images/png/affiliate-banners/ledger_nano_x.png')
    },
    UNSTOPPABLE_DOMAINS: {
        url: 'https://unstoppabledomains.com/r/moonlet',
        image: require('../../assets/images/png/affiliate-banners/unstoppabledomains.png')
    }
};

export interface IAffiliateBannerType {
    url: string;
    image: any;
}
