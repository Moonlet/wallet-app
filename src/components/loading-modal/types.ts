import { TransactionMessageType, TransactionMessageText } from '../../core/blockchain/types';

export interface ILoadingModalMessage {
    type: TransactionMessageType;
    text?: TransactionMessageText;
    component?: any; // React.ComponentType<any>;
}

export interface ILoadingModalState {
    isVisible: boolean;
    message: ILoadingModalMessage;
}
