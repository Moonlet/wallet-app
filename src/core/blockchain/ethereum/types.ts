export enum MethodSignature {
    INCREASE_ALLOWANCE = 'increaseAllowance(address,uint256)',
    DELEGATE = 'delegate(address,uint256)',
    UNDELEGATE = 'undelegate(address,uint256)',
    WITHDRAW_DELEGATED = 'withdrawDelegated(address,address)',
    TRANSFER = 'transfer(address,uint256)'
}
