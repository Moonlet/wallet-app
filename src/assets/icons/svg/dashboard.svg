<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 22.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 24 24" style="enable-background:new 0 0 24 24;" xml:space="preserve">
<title>dashboard</title>
<g>
	<path fill-rule="evenodd" clip-rule="evenodd" d="M12.75 14.4382C12.7509 14.888 12.3873 15.2535 11.9375 15.2549H5.4375C5.22143 15.2544 5.01439 15.1681 4.86199 15.0149C4.7096 14.8618 4.62434 14.6543 4.625 14.4382V5.44655C4.62456 5.2304 4.71 5.02292 4.86253 4.86977C5.01506 4.71661 5.22219 4.63032 5.43834 4.62988L11.9383 4.64072C12.3878 4.64255 12.7509 5.00792 12.75 5.45738V14.4382Z" stroke="#9A99A2" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M23.3717 9.45324C23.3685 9.89888 23.0048 10.2577 22.5592 10.2549H16.0592C15.8451 10.2565 15.6392 10.1728 15.4868 10.0225C15.3344 9.87211 15.248 9.66733 15.2467 9.45324V5.4424C15.2494 4.99722 15.6123 4.63843 16.0575 4.64074L22.5575 4.6299C23.0036 4.62667 23.368 4.98546 23.3717 5.43157V9.45324Z" stroke="#9A99A2" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M12.75 22.5757C12.7487 22.7903 12.6622 22.9956 12.5095 23.1464C12.3568 23.2973 12.1504 23.3812 11.9358 23.3799L5.43583 23.3691C4.99019 23.3709 4.6273 23.0114 4.625 22.5657V18.5582C4.6261 18.3439 4.71235 18.1388 4.86475 17.9882C5.01715 17.8375 5.22319 17.7536 5.4375 17.7549H11.9375C12.3836 17.7526 12.7472 18.1121 12.75 18.5582V22.5757Z" stroke="#9A99A2" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M16.06 23.3799C15.8437 23.3794 15.6365 23.293 15.4839 23.1397C15.3314 22.9864 15.246 22.7787 15.2467 22.5624V13.5724C15.246 13.3562 15.3312 13.1487 15.4836 12.9954C15.636 12.842 15.843 12.7555 16.0592 12.7549H22.5592C22.7753 12.7555 22.9824 12.842 23.1347 12.9954C23.2871 13.1487 23.3723 13.3562 23.3717 13.5724V22.5516C23.3726 23.0012 23.0096 23.3668 22.56 23.3691L16.06 23.3799Z" stroke="#9A99A2" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
</g>
</svg>