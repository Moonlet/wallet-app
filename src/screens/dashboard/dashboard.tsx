import React from 'react';
import { View, Animated, TouchableOpacity, Platform, ScrollView } from 'react-native';
import { Text } from '../../library';
import { INavigationProps } from '../../navigation/with-navigation-params';
import { TokenDashboard } from '../../components/token-dashboard/token-dashboard';
import { AccountCreate } from '../../components/account-create/account-create';
import { IReduxState } from '../../redux/state';
import { IWalletState, IAccountState } from '../../redux/wallets/state';
import { Blockchain, ChainIdType } from '../../core/blockchain/types';
import LinearGradient from 'react-native-linear-gradient';

import stylesProvider from './styles';
import { smartConnect } from '../../core/utils/smart-connect';
import { connect } from 'react-redux';
import { withTheme, IThemeProps } from '../../core/theme/with-theme';
import { getBalance, setSelectedBlockchain } from '../../redux/wallets/actions';
import { getBlockchain } from '../../core/blockchain/blockchain-factory';
import {
    getSelectedWallet,
    getSelectedAccount,
    getSelectedBlockchain,
    getSelectedBlockchainAccounts
} from '../../redux/wallets/selectors';
import { HeaderIcon } from '../../components/header-icon/header-icon';
import { Icon } from '../../components/icon';
import { themes } from '../../navigation/navigation';
import {
    ICON_SIZE,
    ICON_CONTAINER_SIZE,
    BASE_DIMENSION,
    normalize,
    SCREEN_WIDTH
} from '../../styles/dimensions';
import { ConnectExtensionWeb } from '../../core/connect-extension/connect-extension-web';
import { openBottomSheet } from '../../redux/ui/bottomSheet/actions';
import { BottomSheetType } from '../../redux/ui/bottomSheet/state';
import { calculateBalance } from '../../core/utils/balance';
import { getBlockchains, getChainId } from '../../redux/preferences/selectors';
import { NavigationEvents, StackActions } from 'react-navigation';
import { TestnetBadge } from '../../components/testnet-badge/testnet-badge';
import { IExchangeRates } from '../../redux/market/state';
import { formatAddress } from '../../core/utils/format-address';
import { Amount } from '../../components/amount/amount';
import { WalletType } from '../../core/wallet/types';
import { LoadingIndicator } from '../../components/loading-indicator/loading-indicator';

const ANIMATION_MAX_HEIGHT = normalize(160);
const ANIMATION_MIN_HEIGHT = normalize(70);

export interface IReduxProps {
    wallet: IWalletState;
    walletsNr: number;
    getBalance: typeof getBalance;
    blockchains: Blockchain[];
    openBottomSheet: typeof openBottomSheet;
    selectedAccount: IAccountState;
    selectedBlockchain: Blockchain;
    exchangeRates: IExchangeRates;
    setSelectedBlockchain: typeof setSelectedBlockchain;
    isCreateAccount: boolean;
    selectedBlockchainAccounts: IAccountState[];
    userCurrency: string;
    chainId: ChainIdType;
}

const mapStateToProps = (state: IReduxState) => {
    const selectedAccount = getSelectedAccount(state);

    return {
        wallet: getSelectedWallet(state),
        walletsNr: Object.keys(state.wallets).length,
        blockchains: getBlockchains(state),
        selectedBlockchain: getSelectedBlockchain(state),
        selectedAccount: getSelectedAccount(state),
        exchangeRates: state.market.exchangeRates,
        isCreateAccount: state.ui.screens.dashboard.isCreateAccount,
        selectedBlockchainAccounts: getSelectedBlockchainAccounts(state),
        userCurrency: state.preferences.currency,
        chainId: selectedAccount ? getChainId(state, selectedAccount.blockchain) : ''
    };
};

const mapDispatchToProps = {
    getBalance,
    openBottomSheet,
    setSelectedBlockchain
};

interface IState {
    extraSelectedBlockchain: Blockchain;
    isLoading: boolean;
}

const MyTitle = ({ text }) => (
    <Text
        style={{
            flex: 1,
            fontSize: normalize(20),
            lineHeight: normalize(25),
            letterSpacing: 0.38,
            textAlign: 'center'
        }}
    >
        {text}
    </Text>
);
const MyConnectedTitle = connect((state: IReduxState) => ({
    text: (getSelectedWallet(state) || {}).name
}))(MyTitle);

const navigationOptions = ({ navigation }: any) => ({
    headerTitle: () => <MyConnectedTitle />,
    headerLeft: <HeaderIcon />,
    headerRight: (
        <View style={{ flexDirection: 'row' }}>
            <TouchableOpacity
                style={{ width: ICON_CONTAINER_SIZE }}
                onPress={() => navigation.navigate('Wallets')}
            >
                <Icon
                    name="money-wallet-1"
                    size={ICON_SIZE}
                    style={{ color: themes.dark.colors.accent }}
                />
            </TouchableOpacity>
            <TouchableOpacity
                style={{ width: ICON_CONTAINER_SIZE }}
                onPress={() => navigation.state.params.setDashboardMenuBottomSheet()}
            >
                <Icon
                    name="navigation-menu-vertical"
                    size={ICON_SIZE}
                    style={{ color: themes.dark.colors.accent }}
                />
            </TouchableOpacity>
        </View>
    )
});

export class DashboardScreenComponent extends React.Component<
    INavigationProps & IReduxProps & IThemeProps<ReturnType<typeof stylesProvider>>,
    IState
> {
    public static navigationOptions = navigationOptions;
    private animationValue: any = new Animated.Value(0);

    constructor(
        props: INavigationProps & IReduxProps & IThemeProps<ReturnType<typeof stylesProvider>>
    ) {
        super(props);
        this.state = {
            extraSelectedBlockchain: undefined,
            isLoading: false
        };
    }

    public async componentDidMount() {
        if (Platform.OS === 'web') {
            this.setState({ isLoading: true });
            if ((await ConnectExtensionWeb.isConnected()) === false) {
                this.setState({ isLoading: false });
                this.props.navigation.navigate('OnboardingNavigation');
            } else {
                this.setState({ isLoading: false });
                ConnectExtensionWeb.listenLastSync();
            }
        } else {
            if (this.props.blockchains.length === 0 || this.props.walletsNr < 1) {
                // maybe check this in another screen?
                this.props.navigation.dispatch(StackActions.popToTop());
                this.props.navigation.navigate('OnboardingNavigation');
            }
        }

        this.props.navigation.setParams({
            setDashboardMenuBottomSheet: this.setDashboardMenuBottomSheet
        });
    }

    public componentDidUpdate(prevProps: IReduxProps) {
        if (this.props.selectedBlockchain !== prevProps.selectedBlockchain) {
            const blockchainNotFound =
                this.props.blockchains.slice(0, 4).indexOf(this.props.selectedBlockchain) === -1;

            if (blockchainNotFound) {
                this.setState({ extraSelectedBlockchain: this.props.selectedBlockchain });
            }
        }
    }

    public setDashboardMenuBottomSheet = () => {
        this.props.openBottomSheet(BottomSheetType.DASHBOARD_MENU);
    };

    private renderBlockchain(blockchain: Blockchain) {
        const { styles, blockchains } = this.props;

        return (
            <TouchableOpacity
                key={blockchain}
                style={[
                    styles.blockchainButton,
                    this.props.selectedBlockchain === blockchain && styles.blockchainButtonActive,
                    { width: blockchains.length > 4 ? SCREEN_WIDTH / 4 : 0 }
                ]}
                onPress={() => this.props.setSelectedBlockchain(blockchain)}
            >
                <Text
                    style={
                        this.props.selectedBlockchain === blockchain &&
                        styles.blockchainButtonTextActive
                    }
                >
                    {blockchain && getBlockchain(blockchain).config.ui.blockchainDisplay}
                </Text>
            </TouchableOpacity>
        );
    }

    public renderBottomBlockchainNav() {
        const { styles, blockchains } = this.props;
        const { extraSelectedBlockchain } = this.state;

        return (
            <LinearGradient
                colors={this.props.theme.shadowGradient}
                locations={[0, 0.5]}
                style={styles.selectorGradientContainer}
            >
                <View style={styles.blockchainSelectorContainer} testID="blockchainSelector">
                    <View style={styles.bottomBlockchainContainer}>
                        {blockchains
                            .slice(0, 4)
                            .map(blockchain => this.renderBlockchain(blockchain))}

                        {extraSelectedBlockchain !== undefined &&
                            this.renderBlockchain(extraSelectedBlockchain)}

                        {blockchains.length > 4 && (
                            <TouchableOpacity
                                onPress={() =>
                                    this.props.openBottomSheet(
                                        BottomSheetType.BLOCKCHAIN_NAVIGATION
                                    )
                                }
                                style={styles.expandIconContainer}
                            >
                                <Icon
                                    name="expand"
                                    size={normalize(28)}
                                    style={styles.expandIcon}
                                />
                            </TouchableOpacity>
                        )}
                    </View>
                </View>
            </LinearGradient>
        );
    }

    public onFocus() {
        if (this.props.selectedAccount) {
            this.props.getBalance(
                this.props.selectedAccount.blockchain,
                this.props.selectedAccount.address,
                undefined,
                true
            );
        }
    }

    private renderCoinBalanceCard() {
        const { styles, selectedAccount, chainId } = this.props;
        const blockchain: Blockchain = this.props.selectedBlockchain;

        const balance =
            selectedAccount && calculateBalance(selectedAccount, chainId, this.props.exchangeRates);
        const config = blockchain && getBlockchain(blockchain).config;

        const animatePrimaryAmountFontSize = this.animationValue.interpolate({
            inputRange: [0, ANIMATION_MAX_HEIGHT, ANIMATION_MAX_HEIGHT + 1],
            outputRange: [normalize(30), normalize(19), normalize(19)],
            extrapolate: 'clamp'
        });

        const animateConvertedAmountFontSize = this.animationValue.interpolate({
            inputRange: [0, ANIMATION_MAX_HEIGHT, ANIMATION_MAX_HEIGHT + 1],
            outputRange: [normalize(16), normalize(13), normalize(13)],
            extrapolate: 'clamp'
        });

        const animateCoinBalanceCardHeight = this.animationValue.interpolate({
            inputRange: [0, ANIMATION_MAX_HEIGHT, ANIMATION_MAX_HEIGHT + 1],
            outputRange: [ANIMATION_MAX_HEIGHT, ANIMATION_MIN_HEIGHT, ANIMATION_MIN_HEIGHT],
            extrapolate: 'clamp'
        });

        const animateHideAccountAddress = this.animationValue.interpolate({
            inputRange: [0, ANIMATION_MAX_HEIGHT / 2, ANIMATION_MAX_HEIGHT],
            outputRange: Platform.select({
                default: [0, 0, 1],
                web: [1, 1, 0]
            }),
            extrapolate: 'clamp'
        });

        const animateOpacityAccountAddressWeb = this.animationValue.interpolate({
            inputRange: [0, ANIMATION_MIN_HEIGHT / 2, ANIMATION_MAX_HEIGHT],
            outputRange: [1, 0.5, 0],
            extrapolate: 'clamp'
        });

        const animateParimaryAmountVerticalPadding = this.animationValue.interpolate({
            inputRange: [0, ANIMATION_MAX_HEIGHT / 2, ANIMATION_MAX_HEIGHT],
            outputRange: [BASE_DIMENSION, BASE_DIMENSION / 2, 0],
            extrapolate: 'clamp'
        });

        return (
            <Animated.View
                style={[styles.coinBalanceCard, { height: animateCoinBalanceCardHeight }]}
            >
                <TouchableOpacity
                    onPress={() =>
                        this.props.openBottomSheet(BottomSheetType.ACCOUNTS, { blockchain })
                    }
                >
                    {selectedAccount && (
                        <Animated.View
                            style={[
                                styles.row,
                                {
                                    flex: animateHideAccountAddress,
                                    opacity: Platform.select({
                                        default: 1,
                                        web: animateOpacityAccountAddressWeb
                                    })
                                }
                            ]}
                        >
                            <Text style={styles.account}>
                                {selectedAccount.name || `Account ${selectedAccount.index + 1}`}
                            </Text>
                            <Text style={styles.address}>
                                {formatAddress(selectedAccount.address, blockchain)}
                            </Text>
                        </Animated.View>
                    )}

                    <View style={styles.row}>
                        <Amount
                            style={[
                                styles.mainText,
                                {
                                    fontSize: animatePrimaryAmountFontSize,
                                    paddingVertical: animateParimaryAmountVerticalPadding
                                }
                            ]}
                            amount={String(balance)}
                            token={config.coin}
                            tokenDecimals={config.tokens[config.coin].decimals}
                            blockchain={blockchain}
                            isAnimated={true}
                        />
                        <Icon name="chevron-down" size={normalize(18)} style={styles.icon} />
                    </View>
                    <View style={styles.row}>
                        <Amount
                            style={[
                                styles.secondaryText,
                                { fontSize: animateConvertedAmountFontSize }
                            ]}
                            amount={String(balance)}
                            token={config.coin}
                            tokenDecimals={config.tokens[config.coin].decimals}
                            blockchain={blockchain}
                            convert
                            isAnimated={true}
                        />
                    </View>
                </TouchableOpacity>
            </Animated.View>
        );
    }

    private renderTokenDashboard() {
        const styles = this.props.styles;
        const { blockchains, chainId } = this.props;
        const blockchain: Blockchain = this.props.selectedBlockchain;

        return (
            <View style={{ flex: 1 }}>
                <ScrollView
                    contentContainerStyle={[
                        styles.dashboardContainer,
                        {
                            paddingTop: ANIMATION_MAX_HEIGHT // TODO: animate this
                        }
                    ]}
                    showsVerticalScrollIndicator={false}
                    scrollEventThrottle={16}
                    onScroll={Animated.event([
                        { nativeEvent: { contentOffset: { y: this.animationValue } } }
                    ])}
                    alwaysBounceVertical={false}
                >
                    <TokenDashboard
                        account={this.props.selectedAccount}
                        blockchain={blockchain}
                        navigation={this.props.navigation}
                        showBottomPadding={blockchains?.length > 1}
                        chainId={chainId}
                    />
                </ScrollView>

                {this.renderCoinBalanceCard()}
            </View>
        );
    }

    public render() {
        const { blockchains, styles } = this.props;
        const blockchain: Blockchain = this.props.selectedBlockchain;
        const showCreateAccount =
            this.props.isCreateAccount && this.props.selectedBlockchainAccounts?.length === 0;

        /* Hardware wallets can have only one blockchain active */
        let renderBottomNav = false;
        const isHWWallet = this.props.wallet ? this.props.wallet.type === WalletType.HW : false;
        if (blockchains.length > 1 && !isHWWallet) renderBottomNav = true;

        if (Platform.OS === 'web' && this.state.isLoading) {
            return (
                <View style={styles.container}>
                    <LoadingIndicator />
                </View>
            );
        }

        return (
            <View style={styles.container}>
                <TestnetBadge />

                <NavigationEvents onWillFocus={payload => this.onFocus()} />

                {showCreateAccount && (
                    <AccountCreate blockchain={blockchain} navigation={this.props.navigation} />
                )}

                {!showCreateAccount && this.renderTokenDashboard()}

                {renderBottomNav && this.renderBottomBlockchainNav()}
            </View>
        );
    }
}

export const DashboardScreen = smartConnect(DashboardScreenComponent, [
    connect(mapStateToProps, mapDispatchToProps),
    withTheme(stylesProvider)
]);
