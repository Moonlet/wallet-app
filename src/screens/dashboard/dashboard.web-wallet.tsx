import React from 'react';
import { View } from 'react-native';
import { Text } from '../../library';
import { INavigationProps } from '../../navigation/with-navigation-params';
import stylesProvider from './styles.web-wallet';
import { smartConnect } from '../../core/utils/smart-connect';
import { withTheme, IThemeProps } from '../../core/theme/with-theme';

export class DashboardScreenComponent extends React.Component<
    INavigationProps & IThemeProps<ReturnType<typeof stylesProvider>>
> {
    public render() {
        const { styles } = this.props;

        return (
            <View style={styles.container}>
                <Text>{`Welcome to Moonlet Web Wallet!`}</Text>
            </View>
        );
    }
}

export const DashboardScreen = smartConnect(DashboardScreenComponent, [withTheme(stylesProvider)]);
