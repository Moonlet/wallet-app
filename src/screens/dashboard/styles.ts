import { StyleSheet, Platform } from 'react-native';
import { ITheme } from '../../core/theme/itheme';
import { BASE_DIMENSION } from '../../styles/dimensions';

export default (theme: ITheme) =>
    StyleSheet.create({
        container: {
            flex: 1,
            backgroundColor: theme.colors.appBackground,
            height: Platform.OS === 'web' ? 'calc(100vh - 122px)' : 'auto'
        },
        coinBalanceCard: {
            paddingTop: BASE_DIMENSION * 4,
            paddingBottom: BASE_DIMENSION * 6
        },
        dashboardContainer: {
            flex: 1,
            paddingHorizontal: BASE_DIMENSION * 2
        },
        selectorGradientContainer: {
            position: 'absolute',
            flex: 1,
            left: 0,
            right: 0,
            height: 100,
            bottom: 0,
            zIndex: 1,
            justifyContent: 'flex-end'
        },
        blockchainSelectorContainer: {
            backgroundColor: theme.colors.cardBackground,
            borderRadius: 20,
            height: 40,
            flexDirection: 'row',
            marginBottom: 12,
            marginHorizontal: BASE_DIMENSION * 2,
            overflow: 'hidden'
        },
        blockchainButton: {
            flex: 1,
            flexBasis: 0,
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 16,
            margin: BASE_DIMENSION / 2
        },
        blockchainButtonActive: {
            backgroundColor: theme.colors.textTertiary
        },
        blockchainButtonTextActive: {
            fontSize: 16,
            lineHeight: 18,
            color: theme.colors.accent
        },
        coinDashboard: {
            flex: 1,
            alignSelf: 'stretch'
        },
        expandIconContainer: {
            marginRight: BASE_DIMENSION * 2,
            justifyContent: 'center'
        },
        expandIcon: {
            color: theme.colors.accent,
            alignSelf: 'center'
        },
        bottomBlockchainContainer: {
            flex: 1,
            flexDirection: 'row'
        }
    });
