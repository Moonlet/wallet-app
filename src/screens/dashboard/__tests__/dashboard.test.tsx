import React from 'react';
import { Dimensions } from 'react-native';
import { DashboardScreenComponent, IProps, IReduxProps } from '../dashboard';
import { WalletType } from '../../../core/wallet/types';
import { Blockchain } from '../../../core/blockchain/types';
import { darkTheme } from '../../../styles/themes/dark-theme';
import styleProvider from '../styles';

import { shallow } from 'enzyme';

const props: IProps & IReduxProps = {
    // @ts-ignore
    navigation: {
        navigate: jest.fn()
    },
    wallet: {
        id: 'walletId',
        type: WalletType.HD,
        accounts: [
            {
                index: 1,
                blockchain: Blockchain.ZILLIQA,
                address: 'zil1vs74hw5k21233h432kj321l3k21b',
                publicKey: '1',
                balance: 12332
            },
            {
                index: 2,
                blockchain: Blockchain.ZILLIQA,
                address: 'zil1vs74hw5k21233h432kj321l3k21b',
                publicKey: '1',
                balance: 3432
            },
            {
                index: 8,
                blockchain: Blockchain.ETHEREUM,
                address: '0xeAE3Dcc2E37AD412312ASds2d4a6065A831eF89E',
                publicKey: '1',
                balance: 3.5
            },
            {
                index: 9,
                blockchain: Blockchain.COSMOS,
                address: 'cosmos123ksdadasda',
                publicKey: '1',
                balance: 220
            },
            {
                index: 8,
                blockchain: Blockchain.STELLAR,
                address: 'STLsadlij23lj313',
                publicKey: '1',
                balance: 1234
            }
        ]
    },
    styles: styleProvider(darkTheme),
    theme: darkTheme
};

jest.useFakeTimers();

describe('dashboard screen component', () => {
    beforeAll(() => {
        jest.mock('react-native/Libraries/Animated/src/Animated.js', () => {
            const ActualAnimated = jest.requireActual(
                'react-native/Libraries/Animated/src/Animated.js'
            );
            return {
                ...ActualAnimated,
                timing: (value: any, config: any) => {
                    return {
                        start: (callback: any) => {
                            value.setValue(config.toValue);
                            callback && callback();
                        }
                    };
                }
            };
        });
    });

    it('renders correctly', () => {
        const wrapper = shallow(<DashboardScreenComponent {...props} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('handles scroll correctly', () => {
        const screenWidth = Dimensions.get('window').width;
        const spySetState = jest.spyOn(DashboardScreenComponent.prototype, 'setState');

        const wrapper = shallow(<DashboardScreenComponent {...props} />);
        const instance = wrapper.instance();
        // @ts-ignore
        instance.handleScrollEnd({ nativeEvent: { contentOffset: { x: 0 } } });

        // dont call set state when its the same coin index
        expect(spySetState).not.toHaveBeenCalled();

        // @ts-ignore
        instance.handleScrollEnd({
            nativeEvent: { contentOffset: { x: Math.round(screenWidth * 0.5) } }
        });
        expect(spySetState).toHaveBeenCalledWith({
            coinIndex: 1
        });
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('swtiches to correct coin on press', () => {
        const wrapper = shallow(<DashboardScreenComponent {...props} />);
        const selectionButton = wrapper.find('[testID="blockchainSelector"]').childAt(2);
        selectionButton.simulate('press');
        // @ts-ignore
        expect(wrapper.state().coinIndex).toBe(2);
        expect(wrapper.debug()).toMatchSnapshot();
    });
});
