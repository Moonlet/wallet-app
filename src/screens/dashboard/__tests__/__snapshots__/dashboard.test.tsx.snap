// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard screen component renders correctly 1`] = `
"<View style={{...}}>
  <View style={{...}}>
    <ScrollViewMock onMomentumScrollEnd={[Function]} horizontal={true} disableIntervalMomentum={true} overScrollMode=\\"never\\" centerContent={true} snapToAlignment=\\"start\\" snapToInterval={375} contentContainerStyle={{...}} showsHorizontalScrollIndicator={false} snapToStart={false} snapToEnd={false} decelerationRate={0.8}>
      <View style={{...}} />
      <withTheme(CoinBalanceCardComponent) balance={{...}} blockchain=\\"ZILLIQA\\" currency=\\"ZIL\\" width={375} toCurrency=\\"USD\\" active={true} />
      <withTheme(CoinBalanceCardComponent) balance={{...}} blockchain=\\"ETHEREUM\\" currency=\\"ETH\\" width={375} toCurrency=\\"USD\\" active={false} />
      <View style={{...}} />
    </ScrollViewMock>
  </View>
  <AnimatedComponent style={{...}}>
    <withTheme(CoinDashboardComponent) accounts={{...}} blockchain=\\"ZILLIQA\\" navigation={{...}} />
  </AnimatedComponent>
  <LinearGradient colors={{...}} locations={{...}} style={{...}} start={{...}} end={{...}}>
    <View style={{...}} testID=\\"blockchainSelector\\">
      <ScrollViewMock horizontal={true} disableIntervalMomentum={true} overScrollMode=\\"never\\" centerContent={true} snapToAlignment=\\"start\\" contentContainerStyle={{...}} showsHorizontalScrollIndicator={false} snapToStart={false} snapToEnd={false} decelerationRate={0.8}>
        <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
          <withTheme(TextComponent) style={{...}}>
            ZIL
          </withTheme(TextComponent)>
        </TouchableOpacity>
        <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
          <withTheme(TextComponent) style={false}>
            ETH
          </withTheme(TextComponent)>
        </TouchableOpacity>
      </ScrollViewMock>
    </View>
  </LinearGradient>
  <withTheme(PasswordModalComponent) obRef={[Function: obRef]} />
</View>"
`;
